Copter, Sub: consolidate set-home functions

rmackay9 opened the pull request:
This PR does two things but only one of them is a functional change:

- reduce the number of set-home-ish methods by adding a "lock" argument.  this is a non-functional change.
- set the compass's initial location using the home location instead of the current GPS location.

This second change resolves a problem in cases where we are not using the GPS (i.e. when using optical flow or visual-odometry).  We resolve this by using the home location instead of the current GPS location.  Now, there is a reasonable argument that we should instead use the ahrs.get_position() (i.e. the vehicle's current position) instead of the home position although they are normally close together.  If people feel it would be best, I could change it to the current location (according to the EKF).

OXINARF commented:
YES! I hated all those home methods, every time I looked at code that involved I went nuts having to look at all of them. So I'm very happy to see this!

Regarding compass, I agree with not using GPS location, but I think we should use AHRS position, simply because you can set home to somewhere away from your actual location (probably not too far away that it matters but still).

jaxxzer commented:
Looks good, thanks Randy.

rmackay9 commented:
Ok, thanks for the feedback.
I've moved compass.set_initial_location to the compass_accumulate method and it uses the vehicle's position now instead of home.  Also rebased on master.

rmackay9 commented:
Thanks for the reviews, merged!
